### PR TITLE
[mlir][spirv] Handle failed conversions of struct elements

### PR DESCRIPTION
**Pull request by Pierre-vh**
LLVMStructTypes could be emitted with some null elements. This caused a crash later in the LLVMDialect verifier.

Now, properly check that all struct elements were successfully converted before passing them to the LLVMStructType ctor.

See #59990

### CODE REVIEW

File: mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp [hunk 1]
```diff
@@ -207,21 +219,19 @@ convertStructTypeWithOffset(spirv::StructType type,
   if (type != VulkanLayoutUtils::decorateType(type))
     return std::nullopt;
 
-  auto elementsVector = llvm::to_vector<8>(
-      llvm::map_range(type.getElementTypes(), [&](Type elementType) {
-        return converter.convertType(elementType);
-      }));
+  SmallVector<Type> elementsVector;
+  if (!convertTypes(converter, type.getElementTypes(), elementsVector))
```
Comment (victor-eds) [inline mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp:223]: Please use [`TypeConverter::convertTypes`](https://mlir.llvm.org/doxygen/classmlir_1_1TypeConverter.html#ad8a2339e9128e1b05df14b075f6c3914) instead

File: mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp [hunk 2]
```diff
@@ -207,21 +219,19 @@ convertStructTypeWithOffset(spirv::StructType type,
...
-static Type convertStructTypePacked(spirv::StructType type,
-                                    LLVMTypeConverter &converter) {
-  auto elementsVector = llvm::to_vector<8>(
-      llvm::map_range(type.getElementTypes(), [&](Type elementType) {
-        return converter.convertType(elementType);
-      }));
+static std::optional<Type>
+convertStructTypePacked(spirv::StructType type, LLVMTypeConverter &converter) {
+  SmallVector<Type> elementsVector;
+  if (!convertTypes(converter, type.getElementTypes(), elementsVector))
```
Comment (victor-eds) [inline mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp:233]: Same as above

**Review by victor-eds — COMMENTED** [review]
This is a nice catch. I see `convertArrayType`, `convertPointerType` and `convertRuntimeArrayType` have the same issue. Can you please also address in this PR?

**Review by kuhar — CHANGES_REQUESTED** [review]
+1 for using TypeConverter

**Review by victor-eds — COMMENTED** [review]
Change LGTM. Also checking element types for other compound types (`ArrayType`, `PointerType` and `RuntimeArrayType`) would be a plus.

**Comment by Pierre-vh** [comment]
> Change LGTM. Also checking element types for other compound types (`ArrayType`, `PointerType` and `RuntimeArrayType`) would be a plus.

I'm not too familiar with SPIRV semantics so I just want to focus on avoiding the crash in this patch.
@kuhar is this good to land?

File: mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp [hunk 3]
```diff
@@ -207,21 +207,19 @@ convertStructTypeWithOffset(spirv::StructType type,
   if (type != VulkanLayoutUtils::decorateType(type))
     return std::nullopt;
 
-  auto elementsVector = llvm::to_vector<8>(
-      llvm::map_range(type.getElementTypes(), [&](Type elementType) {
-        return converter.convertType(elementType);
-      }));
+  SmallVector<Type> elementsVector;
+  if (converter.convertTypes(type.getElementTypes(), elementsVector).failed())
```
Comment (kuhar) [inline mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp:211]: Prefer `if (failed(...))`

**Review by kuhar — COMMENTED** [review]
Looks good % the failure checks

**Review by kuhar — APPROVED** [review]
LGTM!

File: mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp [hunk 4]
```diff
@@ -207,21 +207,19 @@ convertStructTypeWithOffset(spirv::StructType type,
...
 
 /// Converts SPIR-V struct with no offset to packed LLVM struct.
-static Type convertStructTypePacked(spirv::StructType type,
-                                    LLVMTypeConverter &converter) {
-  auto elementsVector = llvm::to_vector<8>(
-      llvm::map_range(type.getElementTypes(), [&](Type elementType) {
-        return converter.convertType(elementType);
-      }));
+static std::optional<Type>
+convertStructTypePacked(spirv::StructType type, LLVMTypeConverter &converter) {
```
Comment (kuhar) [inline mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp:219]: I've just noticed this: it's already the convention to return `nullptr` when type conversion failed. Can we drop the `optional` and keep `Type` as the return value?

Comment (victor-eds) [reply mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp:219]: :+1: LGTM when this is done (also above)

**Review by antiagainst — APPROVED** [review]
Nice, thanks for the fix!

**Pull request merged**